crates/walrus-service/src/node.rs

```diff
@@ -2832,54 +3043,71 @@ impl ServiceState for StorageNodeInner {
         &self,
         metadata: UnverifiedBlobMetadataWithId,
+        intent: UploadIntent,
     ) -> Result<bool, StoreMetadataError> {
-        let Some(blob_info) = self
+        let blob_id = *metadata.blob_id();
+
+        let blob_info = self
             .storage
-            .get_blob_info(metadata.blob_id())
-            .context("could not retrieve blob info")?
-        else {
-            return Err(StoreMetadataError::NotCurrentlyRegistered);
-        };
+            .get_blob_info(&blob_id)
+            .context("could not retrieve blob info")?;
 
-        if let Some(event) = blob_info.invalidation_event() {
+        if let Some(event) = blob_info
+            .as_ref()
+            .and_then(|info| info.invalidation_event())
+        {
             return Err(StoreMetadataError::InvalidBlob(event));
         }
 
+        let encoding_type = metadata.metadata().encoding_type();
         ensure!(
-            blob_info.is_registered(self.current_committee_epoch()),
-            StoreMetadataError::NotCurrentlyRegistered,
+            encoding_type.is_supported(),
+            StoreMetadataError::UnsupportedEncodingType(encoding_type),
         );
 
-        if blob_info.is_metadata_stored() {
-            return Ok(false);
-        }
+        let encoding_config = self.encoding_config.clone();
+        let verified = Arc::new(
+            self.thread_pool
+                .clone()
+                .oneshot(move || metadata.verify(&encoding_config))
+                .map(thread_pool::unwrap_or_resume_panic)
+                .await?,
+        );
 
-        // Check if encoding type is supported
-        let encoding_type = metadata.metadata().encoding_type();
-        if !encoding_type.is_supported() {
-            return Err(StoreMetadataError::UnsupportedEncodingType(encoding_type));
+        if blob_info
+            .as_ref()
+            .is_some_and(|info| info.is_registered(self.current_committee_epoch()))
+        {
+            return self.persist_verified_metadata(&blob_id, verified).await;
         }
 
-        let encoding_config = self.encoding_config.clone();
-        let verified_metadata_with_id = self
-            .thread_pool
-            .clone()
-            .oneshot(move || metadata.verify(&encoding_config))
-            .map(thread_pool::unwrap_or_resume_panic)
-            .await?;
+        if !intent.is_pending() {
+            return Err(StoreMetadataError::NotCurrentlyRegistered);
+        }
 
-        self.storage
-            .put_verified_metadata(&verified_metadata_with_id)
+        self.pending_metadata_cache
+            .insert(blob_id, verified)
             .await
-            .context("unable to store metadata")?;
-
-        self.metrics
-            .uploaded_metadata_unencoded_blob_bytes
-            .observe(verified_metadata_with_id.as_ref().unencoded_length() as f64);
-        self.metrics.metadata_stored_total.inc();
+            .map_err(|_| StoreMetadataError::CacheSaturated)?;
```

Comment:
question: this means that someone can send large number of cache requests without registering the blob, and cause other metadata upload to fail?